kentikapi/users_integration_test.go

```diff
@@ -194,6 +195,108 @@ func TestClient_GetAllUsers(t *testing.T) {
 	}
 }
 
+func TestClient_GetUser_WithTimeout(t *testing.T) {
+	tests := []struct {
+		name            string
+		responses       []testutil.HTTPResponse
+		expectedResult  *models.User
+		expectedError   bool
+		expectedTimeout bool
+		// timeout and expected requests are strictly connected. The server sleeps for 1 ms before each response.
+		timeout      *time.Duration
+		expectedReqs int
+	}{
+		{
+			name: "default timeout",
+			responses: []testutil.HTTPResponse{
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: false,
+			expectedReqs:    3,
+		}, {
+			name: "status bad request",
+			responses: []testutil.HTTPResponse{
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: false,
+			timeout:         durationPtr(1 * time.Second),
+			expectedReqs:    1,
+		}, {
+			name: "timeout is longer than the wait for response with retries",
+			responses: []testutil.HTTPResponse{
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: false,
+			timeout:         durationPtr(10 * time.Second),
+			expectedReqs:    3,
+		}, {
+			name: "timeout after 1 request",
+			responses: []testutil.HTTPResponse{
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				testutil.NewErrorHTTPResponse(http.StatusBadGateway),
+				{StatusCode: http.StatusBadRequest, Body: `{"error":"Bad Request"}`},
+			},
+			expectedError:   true,
+			expectedTimeout: true,
+			timeout:         durationPtr(5 * time.Millisecond),
+			expectedReqs:    1,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// arrange
+			h := testutil.NewMultipleResponseSpyHTTPHandlerSleep(t, tt.responses)
+			s := httptest.NewServer(h)
+			defer s.Close()
+
+			c := kentikapi.NewClient(kentikapi.Config{
+				APIURL:    s.URL,
+				AuthEmail: dummyAuthEmail,
+				AuthToken: dummyAuthToken,
+				RetryCfg: httputil.RetryConfig{
+					MinDelay: durationPtr(1 * time.Microsecond),
+					MaxDelay: durationPtr(10 * time.Microsecond),
+				},
+				Timeout: tt.timeout,
+			})
+
+			// act
+			result, err := c.Users.Get(context.Background(), testUserID)
+
+			// assert
+			t.Logf("Got result: %v, err: %v", result, err)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
+
+			if tt.expectedTimeout {
+				assert.True(t, strings.Contains(err.Error(), "context deadline exceeded"))
```

Comment:
"context deadline exceeded" string is an implementation detail of the _context_ package, so we should avoid relying on that, even in tests.
Please use something like `assert.True(t, errors.Is(err, context.DeadlineExceeded))`.

Using errors.As() ensures that test will not break when our implementation wraps the error.
Using context.DeadlineExceeded ensures that test will not break when error message changes.

See: https://go.dev/blog/go1.13-errors

Reply:
Done